package modeldata;

message Vec3 
{
	optional float x							= 1 [default = 0.0];
	optional float y							= 2 [default = 0.0];
	optional float z							= 3 [default = 0.0];
}

message Vec4
{
	optional float x							= 1 [default = 0.0];
	optional float y							= 2 [default = 0.0];
	optional float z							= 3 [default = 0.0];
	optional float w							= 4 [default = 0.0];
}

message Euler 
{
	optional float heading						= 1 [default = 0.0];
	optional float pitch						= 2 [default = 0.0];
	optional float roll							= 3 [default = 0.0];
}

message Matrix4
{
	required Vec4				row0			= 1;
	required Vec4				row1			= 2;
	required Vec4				row2			= 3;
	required Vec4				row3			= 4;
}

message Node
{
	optional string				name			= 1;
	optional Matrix4			transformation	= 2;
	optional uint32				subModel		= 3;
	repeated Node				children		= 4;
	repeated uint32				meshIndex		= 5;
}

message Mesh
{
	optional string				name			= 1;	
	required uint32				materialIndex	= 2;
	optional bytes				vertices		= 3;
	optional bytes				faces			= 4;
	optional bool				externalRef		= 5;
}

message Material
{
	required string				name			= 1;
	optional uint32				contents		= 2;
	optional uint32				surfaceFlags	= 3;
}

message Scene
{
	required Node				rootNode		= 1;
	repeated Mesh				meshes			= 2;
	repeated Material			materials		= 3;
}
